File: bumperbot_controller/src/key_teleop.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>  // Required for fcntl()
#include <iostream>

class KeyboardTeleop : public rclcpp::Node
{
public:
    KeyboardTeleop() : Node("keyboard_teleop_node")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::TwistStamped>("/input_key/cmd_vel_stamped", 10);
        RCLCPP_INFO(this->get_logger(), "Use W/A/S/D to move, Q to stop.");

        timer_ = this->create_wall_timer(std::chrono::milliseconds(100),
                                         std::bind(&KeyboardTeleop::publishCommand, this));
        setNonBlocking();
    }

    ~KeyboardTeleop()
    {
        resetTerminal();
    }

private:
    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    char key_ = 0;

    void setNonBlocking()
    {
        tcgetattr(STDIN_FILENO, &original_termios_);
        struct termios new_settings = original_termios_;
        new_settings.c_lflag &= ~(ICANON | ECHO);
        tcsetattr(STDIN_FILENO, TCSANOW, &new_settings);
        fcntl(STDIN_FILENO, F_SETFL, O_NONBLOCK);
    }

    void resetTerminal()
    {
        tcsetattr(STDIN_FILENO, TCSANOW, &original_termios_);
    }

    void publishCommand()
    {
        geometry_msgs::msg::TwistStamped msg;
        char ch;
        if (read(STDIN_FILENO, &ch, 1) > 0)
        {
            key_ = tolower(ch);
        }

        switch (key_)
        {
        case 'w':
            msg.twist.linear.x = 0.5;
            msg.twist.angular.z = 0.0;
            break;
        case 's':
            msg.twist.linear.x = -0.5;
            msg.twist.angular.z = 0.0;
            break;
        case 'a':
            msg.twist.linear.x = 0.0;
            msg.twist.angular.z = 1.0;
            break;
        case 'd':
            msg.twist.linear.x = 0.0;
            msg.twist.angular.z = -1.0;
            break;
        case 'q':
            msg.twist.linear.x = 0.0;
            msg.twist.angular.z = 0.0;
            break;
        default:
            return; // Skip publishing if no valid key
        }

        msg.header.stamp = this->now();
        msg.header.frame_id = "base_link";  // Set this to your robot's frame

        publisher_->publish(msg);
    }

    struct termios original_termios_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<KeyboardTeleop>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_controller/bumperbot_controller/key_teleop.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped
import sys
import tty
import termios
import select
import time


class KeyboardTeleop(Node):
    def __init__(self):
        super().__init__('keyboard_teleop_node')
        self.publisher_ = self.create_publisher(TwistStamped, '/input_key/cmd_vel_stamped', 10)
        self.get_logger().info("Use W/A/S/D to move, Q to stop.")
        self.timer = self.create_timer(0.1, self.publish_command)
        self.key = None
        self.old_settings = termios.tcgetattr(sys.stdin)
        tty.setcbreak(sys.stdin.fileno())

    def destroy_node(self):
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, self.old_settings)
        super().destroy_node()

    def publish_command(self):
        msg = TwistStamped()
        if self.kbhit():
            ch = sys.stdin.read(1)
            self.key = ch.lower()

        if self.key == 'w':
            msg.twist.linear.x = 0.5
            msg.twist.angular.z = 0.0
        elif self.key == 's':
            msg.twist.linear.x = -0.5
            msg.twist.angular.z = 0.0
        elif self.key == 'a':
            msg.twist.linear.x = 0.0
            msg.twist.angular.z = 1.0
        elif self.key == 'd':
            msg.twist.linear.x = 0.0
            msg.twist.angular.z = -1.0
        elif self.key == 'q':
            msg.twist.linear.x = 0.0
            msg.twist.angular.z = 0.0
        else:
            return

        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'base_link'
        self.publisher_.publish(msg)

    def kbhit(self):
        """Check if keyboard was hit (non-blocking)"""
        dr, _, _ = select.select([sys.stdin], [], [], 0)
        return dr != []


def main(args=None):
    rclpy.init(args=args)
    node = KeyboardTeleop()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
